patches/api/0477-Add-paper-version-util-class.patch

```diff
@@ -0,0 +1,188 @@
+@@ -0,0 +1,98 @@
++package io.papermc.paper.util;
++
++import net.kyori.adventure.key.Key;
++import net.kyori.adventure.util.Services;
++import org.bukkit.UnsafeValues;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * A utility class to get information about the server
++ * Works even before Bukkit is initialized (e.g. on bootstrap)
++ */
++public class ServerInfo {
++    private ServerInfo() {
++        throw new UnsupportedOperationException("This class cannot be instantiated");
++    }
++    private static final ServerInfoProvider provider = Services.service(ServerInfoProvider.class).orElseThrow();
++
++    /**
++     * Get the version of the server
++     * @return the version of the server (e.g. "1.20.4", "1.20.2 Pre-release 2", "23w31a")
++     */
++    public static @NotNull String version() {
++        return provider.version();
++    }
++
++    /**
++     * Get the api version of the server
++     * @return the api version of the server (e.g. "1.20.4-R0.1-SNAPSHOT")
++     */
++    public static @NotNull String apiVersion() {
++        return provider.apiVersion();
++    }
++
++    /**
++     * Get the name of the server
++     * @return the name of the server (e.g. "Paper")
++     */
++    public static @NotNull String serverName() {
++        return provider.serverName();
++    }
++
++    /**
++     * Returns true if the server is a minecraft release version,
++     * false is it's a snapshot, pre-release, etc
++     * @return if the server is on a release version
++     */
++    public static boolean isRelease() {
++        return provider.isRelease();
++    }
++
++    /**
++     * Returns the unsafe values for the server for unsafe version values
++     * @return the unsafe values for the server
++     */
++    @Deprecated
++    @NotNull
++    public static UnsafeValues unsafe() {
++        return provider.unsafe();
++    }
++
++    /**
++     * Checks if the server runs exactly the specified version
++     * <p>
++     * Note: Will always return false on non release versions
++     * @param version the version to check (e.g. 1, 20, 4)
++     * @return true if the server runs exactly the specified version
++     */
++    public static boolean is(int @NotNull... version) {
++        return provider.is(version);
++    }
++
++    /**
++     * Checks if the server runs exactly the specified version
++     * @param version the version to check (e.g. "1.20.4", "1.20.2 Pre-release 2", "1.20.2-pre2", "23w31a")
++     * @return true if the server runs exactly the specified version
++     */
++    public static boolean is(@NotNull String version) {
++        return provider.is(version);
++    }
++
++    /**
++     * Checks if the server runs a version which is after this version or the same
++     * @param version the version to check (e.g. 1, 20, 4)
++     * @return true if the server runs on this version or a newer version
++     */
++    public static boolean isAtLeast(int @NotNull... version) {
```

Comment:
The int varargs locks the `is` method into only being useful for releases.
Given we might want to release dev builds for snapshot, this seems unfortunate.
I think it is better to accept a String here and do the parsing ourselfs.

Reply:
supporting strings here as arguments makes the logic as complex as it was for snapshot server handling, as if we are on 1.20.4, it's harder to say if it is older than random 99w99a, the same difficulty when it was vice versa

I don't think plugins will keep support for multiple snapshots or very old snapshot versions even when paper will release dev builds for snapshots, and if they will keep support for one snapshot version, they have #is(String) method to check for it

Reply:
Yea I agree it'll be complex down the line to support snapshots, the parsing logic and manifest comparisons will be a bit of work but I don't think it's worth locking us out of this method working for snapshot comparison.

Even without snapshot support, int varags for this is a less than ideal parameter as `isAtLeast()` would then be valid.

As said in the general review, we are not currently supporting snapshots and hence reducing this pr to release-only while keeping us open to one day add snapshot support would be the goal.

For this, I think parsing a string passed here should be fine with the restriction that this method only takes stringified release versions, e.g 1.20.6 but not 1.20.6-rc1/24w18a.